src/Symfony/Component/Form/Extension/Validator/ViolationMapper/DynamicMappingRule.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Extension\Validator\ViolationMapper;
+
+use Symfony\Component\Form\Exception\ErrorMappingException;
+use Symfony\Component\Form\FormConfigBuilder;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\PropertyAccess\PropertyPath;
+
+/**
+ * @internal
+ */
+class DynamicMappingRule extends AbstractMappingRule
+{
+    private $propertyPathLength;
+
+    private $matchRules;
+
+    private $isPrefixRules;
+
+    public function __construct(FormInterface $origin, string $propertyPath, string $targetPath)
+    {
+        if (substr_count($propertyPath, '*') !== substr_count($targetPath, '*')) {
+            throw new ErrorMappingException(sprintf('The number of "*" must be equals on both sides for the dynamic mapping rule "%s => %s" in "%s".', $propertyPath, $targetPath, $origin->getName()));
+        }
+
+        parent::__construct($origin, $targetPath);
+
+        $propertyPath = new PropertyPath($propertyPath);
+
+        $this->propertyPathLength = $propertyPath->getLength();
+
+        foreach ($propertyPath->getElements() as $ruleElement) {
+            if ('*' === $ruleElement) {
+                $this->matchRules[] = function (string $element, &$dynamicValues): bool {
+                    $dynamicValues[] = $element;
+
+                    return true;
+                };
+
+                $this->isPrefixRules[] =  function (): bool {
+                    return true;
+                };
+            } elseif (false === strpos($ruleElement, '*')) {
+                $this->matchRules[] = $this->isPrefixRules[] = function (string $element) use ($ruleElement): bool {
+                    return $element === $ruleElement;
+                };
+            } else {
+                $ruleElementPattern = preg_quote($ruleElement, '/');
+
+                $matchPattern = '/'.str_replace('\\*', '('.FormConfigBuilder::VALID_NAME_PATTERN.')', $ruleElementPattern).'/';
```

Comment:
Damn, this is actually bad since the violation property path does not have to be a valid form name at all. I mixed things up. It can be any valid array key or PHP class property. Need to reiterate on this.

Reply:
After more reflexion, I have realized that this is the right behavior. The dynamic part should always be a form child name. So it matches this regex. It can technically be something else if the child name is intentionally splitted in the Model but it doesn't make sense at all so I guess we don't need to overcomplicate things to support something that nobody does.